mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.mailbox.tools.indexer.events.ImpactingMessageEvent;
+import org.apache.mailbox.tools.indexer.registrations.GlobalRegistration;
+import org.apache.mailbox.tools.indexer.registrations.MailboxRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.Lists;
+
+public class ReIndexerPerformer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReIndexerPerformer.class);
+
+    private static final int NO_LIMIT = 0;
+    private static final int SINGLE_MESSAGE = 1;
+
+    private final MailboxManager mailboxManager;
+    private final ListeningMessageSearchIndex messageSearchIndex;
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+
+    @Inject
+    public ReIndexerPerformer(MailboxManager mailboxManager,
+                              ListeningMessageSearchIndex messageSearchIndex,
+                              MailboxSessionMapperFactory mailboxSessionMapperFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageSearchIndex = messageSearchIndex;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    public Task.Result reIndex(MailboxPath path, ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(path.getUser());
+        return reIndex(path, mailboxSession, reprocessingContext);
+    }
+
+    public Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing");
+        LOGGER.info("Starting a full reindex");
+        List<MailboxPath> mailboxPaths = mailboxManager.list(mailboxSession);
+        GlobalRegistration globalRegistration = new GlobalRegistration();
+        mailboxManager.addGlobalListener(globalRegistration, mailboxSession);
+        try {
+            return handleFullReindexingIterations(mailboxPaths, globalRegistration, reprocessingContext);
+        } finally {
+            mailboxManager.removeGlobalListener(globalRegistration, mailboxSession);
+            LOGGER.info("Full reindex finished");
+        }
+    }
+
+    private Task.Result handleFullReindexingIterations(List<MailboxPath> mailboxPaths, GlobalRegistration globalRegistration,
+                                                       ReprocessingContext reprocessingContext) {
+        return mailboxPaths.stream()
+            .map(globalRegistration::getPathToIndex)
+            .flatMap(OptionalUtils::toStream)
+            .map(path -> {
+                try {
+                    return reIndex(path, reprocessingContext);
+                } catch (Throwable e) {
+                    LOGGER.error("Error while proceeding to full reindexing on {}", path, e);
+                    return Task.Result.PARTIAL;
+                }
+            })
```

Comment:
could be on previous line

Reply:
agree, and checkstyle should prevent to do that (with a so niiiiiice array out of bound error)